sql/expression/function/aggregation/group_concat.go

```diff
@@ -0,0 +1,263 @@
+	sf          sql.SortFields
+	separator   string
+	selectExprs []sql.Expression
+	maxLen      int
+	returnType  sql.Type
+}
+
+var _ sql.FunctionExpression = &GroupConcat{}
+var _ sql.Aggregation = &GroupConcat{}
+
+func NewEmptyGroupConcat() sql.Expression {
+	return &GroupConcat{}
+}
+
+func NewGroupConcat(distinct string, orderBy sql.SortFields, separator string, selectExprs []sql.Expression, maxLen int) (*GroupConcat, error) {
+	return &GroupConcat{distinct: distinct, sf: orderBy, separator: separator, selectExprs: selectExprs, maxLen: maxLen}, nil
+}
+
+// NewBuffer creates a new buffer for the aggregation.
+func (g *GroupConcat) NewBuffer() sql.Row {
+	var rows []sql.Row
+	var distinctSet = make(map[string]bool)
+
+	return sql.NewRow(rows, distinctSet)
+}
+
+// Update implements the Aggregation interface.
+func (g *GroupConcat) Update(ctx *sql.Context, buffer, originalRow sql.Row) error {
+	evalRow, retType, err := evalExprs(ctx, g.selectExprs, originalRow)
+	if err != nil {
+		return err
+	}
+
+	g.returnType = retType
+
+	// Skip if this is a null row
+	if evalRow == nil {
+		return nil
+	}
+
+	var v interface{}
+	if retType == sql.Blob {
+		v, err = sql.Blob.Convert(evalRow[0])
+	} else {
+		v, err = sql.LongText.Convert(evalRow[0])
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if v == nil {
+		return nil
+	}
+
+	vs := v.(string)
+
+	// Get the current array of rows and the map
+	rows := buffer[0].([]sql.Row)
+	distinctSet := buffer[1].(map[string]bool)
+
+	// Check if distinct is active if so look at and update our map
+	if g.distinct != "" {
+		// If this value exists go ahead and return nil
+		if _, ok := distinctSet[vs]; ok {
+			return nil
+		} else {
+			distinctSet[vs] = true
+		}
+	}
+
+	// Append the current value to the end of the row. We want to preserve the row's original structure for
+	// for sort ordering in the final step.
+	rows = append(rows, append(originalRow, nil, vs))
+
+	buffer[0] = rows
+	buffer[1] = distinctSet
+
+	return nil
+}
+
+// Merge implements the Aggregation interface.
+func (g *GroupConcat) Merge(ctx *sql.Context, buffer, partial sql.Row) error {
+	return g.Update(ctx, buffer, partial)
+}
+
+// cc: https://dev.mysql.com/doc/refman/8.0/en/aggregate-functions.html#function_group-concat
+func (g *GroupConcat) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	rows := row[0].([]sql.Row)
+
+	if len(rows) == 0 {
+		return nil, nil
+	}
+
+	// Execute the order operation if it exists.
+	if g.sf != nil {
+		sorter := &expression.Sorter{
+			SortFields: g.sf,
+			Rows:       rows,
+			Ctx:        ctx,
+		}
+
+		sort.Stable(sorter)
+		if sorter.LastError != nil {
+			return nil, sorter.LastError
+		}
+	}
+
+	sb := strings.Builder{}
+	for i, row := range rows {
+		lastIdx := len(row) - 1
+		if i == len(rows)-1 {
+			sb.WriteString(row[lastIdx].(string))
+		} else {
+			sb.WriteString(row[lastIdx].(string))
+			sb.WriteString(g.separator)
+		}
+	}
+	ret := sb.String()
+
+	if len(ret) > g.maxLen {
```

Comment:
Want to do this during the loop (and exit early), rather than at the end. Otherwise you blow up memory for no reason